### PR TITLE
Some fixes

### PR DESCRIPTION
**Pull request by nikkolasg**
This branch contains somes fixes I did while I wanted to use protobuf for the encoding part of the network library of cothority. 
+All test working now.
+Using type Constructors consistently
+Added error specific error messages. I had great trouble with encoding / decoding arbitrary interface,only knowing the reflect.Type, and these messages could have saved me a lot of troubles.

Only small changes overall.

### CODE REVIEW

File: encode.go [hunk 1]
```diff
@@ -352,7 +355,7 @@ var bytesType = reflect.TypeOf([]byte{})
 
 func (en *encoder) sliceReflect(key uint64, slval reflect.Value) {
 
-	if slval.Kind() != reflect.Slice {
+	if slval.Kind() != reflect.Slice && slval.Kind() != reflect.Array {
```
Comment (bford) [inline encode.go]: maybe insert a "kind := slval.Kind()" before and test that variable, to avoid calling possibly expensive reflection method Kind() twice in a row unnecessarily.

**Comment by bford** [comment]
Looks good, go ahead!

File: decode.go [hunk 2]
```diff
@@ -89,7 +102,7 @@ func (de *decoder) message(buf []byte, sval reflect.Value) error {
 		// Decode the field's value
 		rem, err := de.value(wiretype, buf, field)
 		if err != nil {
-			return err
+			return fmt.Errorf("FIeldName %s: %v", fields[fieldi].Name, err)
```
Comment (Liamsi) [inline decode.go]: typo: 'FIeldName' should be FieldName

File: decode.go [hunk 3]
```diff
@@ -146,7 +158,7 @@ func (de *decoder) value(wiretype int, buf []byte,
 		buf = buf[n+int(v):]
 
 	default:
-		return nil, errors.New("unknown protobuf wire-type")
+		return nil, fmt.Errorf("unknown protobuf wire-type")
```
Comment (Liamsi) [inline decode.go]: Why do we need `Errorf` here? No, formatting directive

File: decode.go [hunk 4]
```diff
@@ -347,7 +359,17 @@ func (de *decoder) slice(slval reflect.Value, vb []byte) error {
 		wiretype = 1 // Packed 64-bit representation
 
 	case reflect.Uint8: // Unpacked byte-slice
-		slval.SetBytes(vb)
+		if slval.Kind() == reflect.Array {
+			if slval.Len() != len(vb) {
+				panic("Array length != bufer length")
```
Comment (Liamsi) [inline decode.go]: typro: buffer instead of bufer

File: encode.go [hunk 5]
```diff
@@ -49,7 +49,11 @@ func Encode(structPtr interface{}) (bytes []byte, err error) {
 		return nil, nil
 	}
 	en := encoder{}
-	en.message(reflect.ValueOf(structPtr).Elem())
+	val := reflect.ValueOf(structPtr)
+	if val.Kind() != reflect.Ptr {
+		return nil, fmt.Errorf("Encode takes a pointer to struct")
```
Comment (Liamsi) [inline encode.go]: Why `Errorf`?

File: generate_test.go [hunk 6]
```diff
@@ -27,8 +27,9 @@ message test {
   required float f32 = 11;
   required double f64 = 12;
   required bytes bytes = 13;
-  required string string = 14;
-  required emb struct = 15;
+  required bytes array = 14;
+  required string string = 15;
+  required emb struct = 16;
```
Comment (Liamsi) [inline generate_test.go]: are you sure this is what would be generated as a `.proto` file? Look here: https://developers.google.com/protocol-buffers/docs/proto#nested

**Comment by Liamsi** [comment]
Besides the comments, looks very fine to me

File: generate_test.go (hunk 6 above)
Comment (Liamsi) [inline generate_test.go]: More specific: I think the correct 'proto'-file needs sth similar to:
```
message emb {
   required Id int; 
   // ...
}
```
before the usage of `emb` to be a valid proto file

Comment (Liamsi) [inline generate_test.go]: Discussed with Nicolas: this is out of scope for this issue. And there is also another test for handling nested types `TestGeneratePersonExample`

**Comment by ineiti** [comment]
My 2 cents: correct the Errorf and English, then open a new issue regarding the proto-file.
